contributors/design-proposals/network/service-topology-key.md

```diff
@@ -0,0 +1,116 @@
+# topology aware routing of services
+
+Author: @m1093782566
+
+## Objective
+
+Provide a generic way to provide the local-only service route, say "topology aware routing of service". Local means "same topology level", e.g. same node, same rack, same failure zone, same failure region, same cloud provider etc. To clarify, this proposal is intended to resolve the user case that user always want to visit a local service's backend even though for some reason local service's backend is not available.
+
+## GOAL
+
+A generic way to support various topology levels, regarless whatever the topology key is.
+
+## Non-goal
+
+Scheduler spreading to implement this sort of topology guarantee.
+
+## Use cases:
+
+* Logging agents such as fluentd. Deploy fluentd as DaemonSet and applications only need to communicate with the fluentd in the same node.
```

Comment:
its not clear to me how this use case is handled ? This currently requires a per node service .  Would creating a service with topologyKey=kubernetes.io/hostname be enough for this use case ? Applications still use the same service cluster ip, but its guranteed to go through the fluentd pod on the same node or fail if none exist ?

Reply:
May be a good idea to show some examples of the service spec and the use case it satsifies.

Reply:
I updated the API - following "Required/Preferred" in podAffinity, PTAL.

If still unclear, please let me know. Thanks!

Reply:
> Applications still use the same service cluster ip, but its guranteed to go through the fluentd pod on the same node or fail if none exist ?

Scheduler spreading to implement this sort of topology guarantee is non-goal of this proposal.

Reply:
> May be a good idea to show some examples of the service spec and the use case it satsifies.

An example is added, PTAL.